constraint-solver/src/solver.rs

```diff
@@ -96,8 +110,187 @@ pub enum Error {
 /// An assignment of a variable.
 pub type VariableAssignment<T, V> = (V, GroupedExpression<T, V>);
 
+/// We introduce new variables.
```

Comment:
There seems to be some duplication with `constraint-solver/src/boolean_extractor.rs`. Could we move the entire solver there? The `SolverImpl` (should we rename it to `CoreSolver`?) does not need to know the `BooleanExtractedSolver` and could be used without it, so I think they should be in separate files.

Reply:
Yes, I wanted to rename `SolverImpl` to something like `BaseSolver` or `CoreSolver`, but in an upcoming pr. I will clean this all up later if it's ok.